### PR TITLE
Access menu shortcuts bar and some refactor

### PR DESCRIPTION
**Pull request by boholder**
with https://github.com/khanshoaib3/minecraft-access-i18n/pull/47

# Changelog
## Feature Updates
- **BREAKING CHANGES**: Rename **Narrator Menu** aka F4 Menu to Access Menu
- **Access Menu**: Access Menu functions can be triggered with Alt + upper number keys

### CODE REVIEW

**Comment by boholder** [comment]
Test failed, gonna fix it

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/mixin/MinecraftClientMixin.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/mixin/MinecraftClientMixin.java]: This method name is way too long, consider naming it `handleInputEvents` to match where it is being injected into.
```suggestion
    private void handleInputEvents(CallbackInfo ci) {
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/system/KeyUtils.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/system/KeyUtils.java]: This method name is a bit verbose, `isAltPressed` would be better.
```suggestion
    public static boolean isAltPressed() {
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/TimingKeystroke.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/TimingKeystroke.java]: Consider `TimedKeystroke`.
```suggestion
public abstract class TimedKeystroke extends Keystroke {
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/MenuKeystroke.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/MenuKeystroke.java]: Consider using JetBrains' `@Contract` annotation to get warnings for incorrect usage.
```suggestion
    @Contract(pure = true)
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java]: This is not a constant and thus should be in camelCase, not SCREAMING_SNAKE_CASE. `keystrokes` may also be a better name because "existing" doesn't really add any meaning. This field can also be final.
```suggestion
    protected static final List<Keystroke> keystrokes = new ArrayList<>();
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java]: Something like `updateKeystrokes` would be more clear.
```suggestion
    public static void updateKeystrokes() {
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java]: Redundant import, included in wildcard on line 5:
```java
import com.github.khanshoaib3.minecraft_access.features.*;
```
```suggestion
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/config/ConfigMap.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/config/ConfigMap.java]: This is a breaking change and should be documented as such.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java]: `MinecraftClient.getInstance()` is never null, and thus checking for it is superfluous. This is also significantly clearer as a regular null check as opposed to using a variable.
```suggestion
        if (WorldUtils.getClientPlayer() != null) {
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/MouseKeySimulation.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/features/MouseKeySimulation.java]: This is not a constant.
```suggestion
    private static boolean enabled;
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java]: Should the key be renamed to be consistent with everything else?
```suggestion
    private static final Keystroke key = new Keystroke(KeyBindingsHandler.getInstance().healthAndHungerNarrationKey);
```

File: docs/features.md
Comment (sgmedev) [inline docs/features.md]: Should this be renamed to be consistent with the name used elsewhere?
```suggestion
### Health and Hunger
```

File: docs/keybindings.md
Comment (sgmedev) [inline docs/keybindings.md]: Should this be renamed to be consistent with the name used elsewhere?
```suggestion
## Health and Hunger
```

**Review by sgmedev — CHANGES_REQUESTED** [review]
Looks good, just a few minor considerations mainly with naming and a lack of consistency with whether it's named `Health N Hunger` or `Health and Hunger`, I personally prefer the latter. The renaming of `Narrator Menu` config option to `Access Menu` should be included in the changelog as it is a breaking change.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java]: I think if this mod is running on server, that instance is null.

**Comment by boholder** [comment]
@TheSuperGamer20578 perfect review! It exposures many naming issues. Thank you for pointing out them!

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java
Comment (sgmedev) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java]: That would make sense, I kind of only looked at the constructor for `MinecraftClient` and saw that it was always set there forgetting that the constructor is only called from the client. I still feel that the additional nesting that `Objects.nonNull()` adds makes it significantly harder to read, the [Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#nonNull-java.lang.Object-) for it states that it exists for use as a predicate for methods such as `Stream.filter()`. Using a variable for this condition is also superfluous because a regular null check is usually easy enough to read.
```suggestion
        if (MinecraftClient.getInstance() != null && WorldUtils.getClientPlayer() != null) {
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Interval.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Interval.java]: Having a not in a predicate name can be confusing, especially if the result is ever inverted as that would result in a double negative (`!hasNotEnded()`). Something like `isRunning` would be more clear; however, I see that there is already a field with that name. Additionally, the Javadoc states that it will not auto-reset the timer but then `isReady` is called which will auto-reset the timer.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/MouseKeySimulation.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/features/MouseKeySimulation.java]: Should these be different intervals?

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/LockingHandler.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/LockingHandler.java]: Should this not be an else if statement?
```suggestion
            } else if (lastAimAssistCue != 0 || bowState != lastBowState) {
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java]: Typo: `Instants` → `Instances`

I also feel that something along the lines of `updateInstances` or `updateAllStates` would be better because `updateAllInstancesState` is a bit verbose.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/system/KeyUtils.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/system/KeyUtils.java]: A method reference may be better than returning a lambda.
```suggestion
        return Arrays.stream(keyBindings).anyMatch(this::isKeyPressed);
```

**Review by sgmedev — CHANGES_REQUESTED** [review]
Just a few more minor tweaks and then I think it should be ready to merge.

**Comment by sgmedev** [comment]
I’ve also discovered that the new shortcuts trigger on every tick while held and `alt+4` and `alt+5` cause the game to freeze for a good few seconds, which is exacerbated by them triggering many times when you hold them down.

**Comment by boholder** [comment]
> I’ve also discovered that the new shortcuts trigger on every tick while held and `alt+4` and `alt+5` cause the game to freeze for a good few seconds, which is exacerbated by them triggering many times when you hold them down.

Roger, ~~I'll fix it and other refactors in Sunday.~~

20241013 edit: My head isn't clear... I want to delay the fixing for another week.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/MainClass.java]: Yep, you're right, will accept it.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/MouseKeySimulation.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/features/MouseKeySimulation.java]: ~~They are different IntervalKeystroke instances, so...~~ oh I get you

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/LockingHandler.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/LockingHandler.java]: This class is complicate, I already refactored it but it's still complicated, @Molitvan 's bow aiming assist added more complexity. That's why I want to recode the POI and locking system.

**Comment by boholder** [comment]
@TheSuperGamer20578 I suggest you review the new part commit by commit, I followed the "small commit" principal.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java
Comment (sgmedev) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java]: How has this been resolved? I don’t see any changes

Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java]: > How has this been resolved? I don’t see any changes

@TheSuperGamer20578 Now the code is:
https://github.com/khanshoaib3/minecraft-access/pull/349/files#diff-adafc93f481283c43ffb98839e6fca7f44e2ef8e069104f7544b781ffc0e66ccL133
```
    /**
     * Let this method handle state updates,
     * no need to manually call the update method in every feature.
     */
    public static void updateAllInstantsState() {
```

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/access_menu/AccessMenu.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/features/access_menu/AccessMenu.java]: It would be better to add the interval here instead of using an index and separate array; along with alleviating the need to keep multiple arrays in sync, this also eliminates the need for magic numbers as seen on lines +78 and +79. Though I feel that this should be completely refactored so that it only needs to use a single array as opposed to two arrays in seperate classes that must be kept in sync.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/mixin/BookEditScreenMixin.java
Comment (sgmedev) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/mixin/BookEditScreenMixin.java]: Is there supposed to be a blank line here?
```suggestion
```

**Review by sgmedev — CHANGES_REQUESTED** [review]
`IntervalKeystroke` needs some work. My testing immediately found that camera controls are quite flakey as to whether or not they work when keys are tapped and that sometimes double taps are ignored, though the double tap behaviour is also observed on the `1.21` branch.

What should happen is that on key down is:

1. The action should always be executed promptly and unconditionally

2. The interval should be reset to prevent inconsistent timing of the first repetition

When the user taps a key, it is expected that the action will be executed promptly once, if the user double-taps a key they expect the action to be executed promptly a second time.

The current, undesirable behaviour can be observed in this screen recording:

https://github.com/user-attachments/assets/e7674f88-0ae9-49fc-b95e-a51546ccc304

I also found some issues with the access menu implementation; however, this can wait for a later refactor.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java
Comment (sgmedev) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java]: Is that not the same as the code I reviewed earlier and suggested a change on?

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/mixin/BookEditScreenMixin.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/mixin/BookEditScreenMixin.java]: Ah... ignore it, IDEA won't auto format it, won't fix.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/access_menu/AccessMenu.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/features/access_menu/AccessMenu.java]: I guess we can fix it with #364  won't fix in this PR.

**Comment by boholder** [comment]
> `IntervalKeystroke` needs some work.

Will look at it this weekend. And... if I understand clearly, this is an old bug, not produced by this PR, how about start another issue for it and won't fix it in release 1.9.0?

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/condition/Keystroke.java]: > Is that not the same as the code I reviewed earlier and suggested a change on?

I don't know...

**Comment by boholder** [comment]
> `IntervalKeystroke` needs some work.

The real reason for this issue is the class scope "global" `Interval` instances in `CameraControls`, `InventoryControls`, `LockingHandler` that manually triggered by player's keystrokes while having an interval limitation, the interval isn't properly working with keystrokes.
Meanwhile the totally auto features like `POIBlock` won't be affected.
Shoaib's previous fix linked interval and keystrokes, ~~I'll simple replace all the keystrokes to `IntervalKeystroke`~~ it seems that his solution is simplest one.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/config/config_menus/NarratorMenuConfigMenu.java
Comment (BrailleBennett) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/config/config_menus/NarratorMenuConfigMenu.java:1]: This file should be access menu not narrator menu

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java
Comment (BrailleBennett) [inline common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java:1]: Can you roll back health and hunger and related i18n changes? PR #371 adds your changes and more as discussed.

**Review by BrailleBennett — CHANGES_REQUESTED** [review]

**Review by sgmedev — DISMISSED** [review]
LGTM! I noticed that the `IntervalKeystroke` behaviour still isn't optimal, however there is no regression and this can be fixed later.

**Comment by boholder** [comment]
> I noticed that the `IntervalKeystroke` behaviour still isn't optimal

I made the change and reverted it, since there is no existing example for me to test the behavior... not hard, just add relative logic in `updateForNextTick` method.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java:1]: @BrailleBennett Sorry, I can't, it's burred under commits, can't simply revert it.

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/config/config_menus/NarratorMenuConfigMenu.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/config/config_menus/NarratorMenuConfigMenu.java:1]: @BrailleBennett I'm afraid this will be resolved in your PR?...

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java
Comment (sgmedev) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java:1]: You can use `git revert` to revert a commit, this basically just creates a new commit that undoes any changes the specified commit made, otherwise you can merge it into here or wait until it gets merged and then resolve the merge conflicts which will likely effectively revert the changes

File: CONTRIBUTING.md
Comment (sgmedev) [inline CONTRIBUTING.md]: This is not necessary, the URL is already at the end of the file

**Review by sgmedev — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (boholder) [reply CONTRIBUTING.md]: But, it's quoted with [ ]...

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java:1]: So... please refer https://github.com/minecraft-access/minecraft-access/pull/349/commits/e557170467c4f1c26670c3525eace5a70e574bbe for modifying doc, refer https://github.com/minecraft-access/minecraft-access/pull/349/files#diff-9c439a76705ddf2f13f81ed1853e75bfd077a5a2163db1e34bf5f3d3b816d3dcR126 for modifying triggering logic, refer https://github.com/minecraft-access/minecraft-access/pull/349/commits/559c27dda31a05ee0ab9e19f3726ebf259ced313 for utilizing the functions, in #371 
I'll revert my changes.

File: CONTRIBUTING.md
Comment (sgmedev) [reply CONTRIBUTING.md]: In markdown you can use `[text]` and then put `[text]: url` at the end of the file, this is equivalent to `[text](url)`. I generally prefer this syntax because it makes lines significantly shorter and thus greatly inproves readability.

Comment (boholder) [reply CONTRIBUTING.md]: TIL!

**Review by sgmedev — DISMISSED** [review]

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/features/HealthAndHunger.java:1]: @boholder TODO

File: common/src/main/java/com/github/khanshoaib3/minecraft_access/config/config_menus/NarratorMenuConfigMenu.java
Comment (boholder) [reply common/src/main/java/com/github/khanshoaib3/minecraft_access/config/config_menus/NarratorMenuConfigMenu.java:1]: @boholder TODO

**Comment by BrailleBennett** [comment]
Make sure to update the changelog to reflect the new additions this PR does.

**Comment by boholder** [comment]
Holy...

**Review by BrailleBennett — APPROVED** [review]

**Pull request merged**
